src/it/java/servlets/module/lesson/XxeLessonIT.java

```diff
@@ -0,0 +1,144 @@
+	{
+		try
+		{
+			String servletClassName = "XxeLesson";
+			log.debug("Creating " + servletClassName + " Servlet Instance");
+			XxeLesson servlet = new XxeLesson();
+			servlet.init(new MockServletConfig(servletClassName));
+
+			request.setContentType("application/xml");
+
+			//Setup Servlet Parameters and Attributes
+			log.debug("Setting Up Params and Atrributes");
+			//for an XML payload instead of addParameter it's setContent with a byte array
+			request.setContent(xmlEmail);
+			//Adding Correct CSRF Token (Token Submitted)
+			request.addHeader("csrfToken", csrfToken);
+
+			log.debug("Running doPost");
+			servlet.doPost(request, response);
+
+			if(response.getStatus() != expectedResponseCode)
+				fail(servletClassName + " Servlet Returned " + response.getStatus() + " Code. " + expectedResponseCode + " Expected");
+			else
+			{
+				log.debug("302 OK Detected");
+				log.debug(servletClassName + " Successful, returning location retrieved: " + response.getContentAsString());
+				return(response.getContentAsString());
+			}
+		}
+		catch(Exception e)
+		{
+			throw e;
+		}
+		return null;
+	}
+
+	@Test
+	public void testLevelValidAnswerXxeInjection()
+	{
+		String userName = "lessonTester";
+		try
+		{
+			//Verify User Exists in DB
+			GetterTest.verifyTestUser(applicationRoot, userName, userName);
+			//Sign in as Normal User
+			log.debug("Signing in as " + userName + " Through LoginServlet");
+			TestProperties.loginDoPost(log, request, response, userName, userName, null, lang);
+			log.debug("Login Servlet Complete, Getting CSRF Token");
+			if(response.getCookie("token") == null)
+				fail("No CSRF Token Was Returned from Login Servlet");
+			String csrfToken = response.getCookie("token").getValue();
+			if(csrfToken.isEmpty())
+			{
+				String message = "No CSRF token returned from Login Servlet";
+				log.fatal(message);
+				fail(message);
+			}
+			else
+			{
+				request.setCookies(response.getCookies());
+				String xxeString = "<?xml version=\"1.0\" encoding=\"ISO-8859-1\"?>"
+						+ "<!DOCTYPE foo ["
+						+ "<!ELEMENT foo ANY >"
+						+ "<!ENTITY xxe SYSTEM \"file://" + System.getProperty("user.dir")
+						+ "/" + FileInputProperties.readfile(System.getProperty("user.dir")
+						+ "/src/main/resources/fileSystemKeys.properties", "xxe.lesson.file") + "\""
+						+ " >]><foo>&xxe;</foo>";
+
+				//String xxeString = "<?xml version=\"1.0\" encoding=\"ISO-8859-1\"?><email>test@test.com</email>";
+
+				//for XML you must use a byte array (xxeString.getBytes())
+				String servletResponse = getModuleDoPost(xxeString.getBytes(), csrfToken, 302);
+
+				Assert.assertTrue(servletResponse.contains("Marsellus Wallace&#39;s Key"));
+				Assert.assertFalse(servletResponse.contains("You must be getting funky"));
+			}
+		}
+		catch(Exception e)
+		{
+			log.fatal("Could not Complete: " + e.toString());
+			fail("Could not Complete: " + e.toString());
+		}
+	}
+
+
+
+}
```

Comment:
Maybe add a test here to see if the level doesn't work if Dangerous Levels have been disabled?

Reply:
Test has been added